File: _site/python_solutions/316.remove-duplicate-letters.cpp
/*
 * @lc app=leetcode id=316 lang=cpp
 *
 * [316] Remove Duplicate Letters
 *
 * https://leetcode.com/problems/remove-duplicate-letters/description/
 *
 * algorithms
 * Hard (32.54%)
 * Total Accepted:    57.2K
 * Total Submissions: 175.6K
 * Testcase Example:  '"bcabc"'
 *
 * Given a string which contains only lowercase letters, remove duplicate
 * letters so that every letter appear once and only once. You must make sure
 * your result is the smallest in lexicographical order among all possible
 * results.
 *
 * Example 1:
 *
 *
 * Input: "bcabc"
 * Output: "abc"
 *
 *
 * Example 2:
 *
 *
 * Input: "cbacdcbc"
 * Output: "acdb"
 *
 */
using namespace std;
using VI = vector<int>;
using LL = long long;

class Solution {
public:
  string removeDuplicateLetters(string s) {
    // int used[26] = {}, cnt[26] = {};
    VI used(26, 0), cnt(used);
    string res;
    for (char &c : s)
      cnt[c - 'a']++;
    for (char &c : s) {
      --cnt[c - 'a'];
      if (used[c - 'a']++ > 0)
        continue;
      while (!res.empty() && res.back() > c && cnt[res.back() - 'a'] > 0) {
        used[res.back() - 'a'] = 0;
        res.pop_back();
      }
      res.push_back(c);
    }
    return res;
  }
};

static const int _ = []() {
  ios::sync_with_stdio(false);
  cin.tie(NULL);
  return 0;
}();

File: _site/python_solutions/316.remove-duplicate-letters.rb
#
# @lc app=leetcode id=316 lang=ruby
#
# [316] Remove Duplicate Letters
#
# https://leetcode.com/problems/remove-duplicate-letters/description/
#
# algorithms
# Hard (31.54%)
# Total Accepted:    51.3K
# Total Submissions: 161.8K
# Testcase Example:  '"bcabc"'
#
# Given a string which contains only lowercase letters, remove duplicate
# letters so that every letter appear once and only once. You must make sure
# your result is the smallest in lexicographical order among all possible
# results.
#
# Example 1:
#
#
# Input: "bcabc"
# Output: "abc"
#
#
# Example 2:
#
#
# Input: "cbacdcbc"
# Output: "acdb"
#
#
# @param {String} s
# @return {String}
def remove_duplicate_letters(s)
  letters = s.chars.uniq.sort
  letters.each do |l|
    return l + remove_duplicate_letters(s[s.index(l)..-1].gsub(l, '')) if s[s.index(l)..-1].chars.uniq.sort == letters
  end
  ''
end

s = 'cbacdcbc'
s = 'bcabc'
p remove_duplicate_letters(s)
